core/src/worker/workflow/managed_run.rs

```diff
@@ -0,0 +1,602 @@
+    async fn incoming_wft(
+        &mut self,
+        wft: NewIncomingWFT,
+    ) -> Result<Option<ActivationOrAuto>, RunUpdateErr> {
+        let activation = if let Some(h) = wft.history_update {
+            self.wfm.feed_history_from_server(h).await?
+        } else {
+            let r = self.wfm.get_next_activation().await?;
+            if r.jobs.is_empty() {
+                return Err(RunUpdateErr {
+                    source: WFMachinesError::Fatal(format!(
+                        "Machines created for {} with no jobs",
+                        self.wfm.machines.run_id
+                    )),
+                    complete_resp_chan: None,
+                });
+            }
+            r
+        };
+
+        if activation.jobs.is_empty() {
+            if self.wfm.machines.outstanding_local_activity_count() > 0 {
+                // If the activation has no jobs but there are outstanding LAs, we need to restart the
+                // WFT heartbeat.
+                if let Some(ref mut lawait) = self.waiting_on_la {
+                    if lawait.completion_dat.is_some() {
+                        panic!("Should not have completion dat when getting new wft & empty jobs")
+                    }
+                    lawait.heartbeat_timeout_task.abort();
+                    lawait.heartbeat_timeout_task = start_heartbeat_timeout_task(
+                        lawait.hb_chan.clone(),
+                        wft.start_time,
+                        lawait.wft_timeout,
+                    );
+                    // No activation needs to be sent to lang. We just need to wait for another
+                    // heartbeat timeout or LAs to resolve
+                    return Ok(None);
+                } else {
+                    panic!(
+                        "Got a new WFT while there are outstanding local activities, but there \
+                     was no waiting on LA info."
+                    )
+                }
+            } else {
+                return Ok(Some(ActivationOrAuto::Autocomplete {
+                    run_id: self.wfm.machines.run_id.clone(),
+                }));
+            }
+        }
+
+        Ok(Some(ActivationOrAuto::LangActivation(activation)))
+    }
+
+    async fn completion(
+        &mut self,
+        mut completion: RunActivationCompletion,
+        heartbeat_tx: &UnboundedSender<Span>,
+    ) -> Result<(), RunUpdateErr> {
+        let resp_chan = completion
+            .resp_chan
+            .take()
+            .expect("Completion response channel must be populated");
+
+        let outcome = async move {
+            // Send commands from lang into the machines then check if the workflow run
+            // needs another activation and mark it if so
+            self.wfm.push_commands(completion.commands).await?;
+            // Don't bother applying the next task if we're evicting at the end of
+            // this activation
+            if !completion.activation_was_eviction {
```

Comment:
What happens if there's an unprocessed WFT while we're evicting? Is this possible?

Reply:
WFT has to be finished and replied to before eviction will happen

Reply:
So there's no way there could be a new task to apply if `completion.activation_was_eviction`?